### PR TITLE
Add Bankless Leader design

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

Leaders are a bottleneck on performance, since they have to do a lot of work of arranging transactions into well formed entries and shredding and broadcasting the entries to the rest of the network.  This CPU intensive work is competing for resources with the Bank.

#### Summary of Changes

Propose a design in which leaders skip execution of transactions and produce valid entries by only verifying that the transaction can afford to be included in the block.

Fixes #

### CODE REVIEW

File: book/src/proposals/bankless-leader.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+For the duration of the block, the leader maintains a temporary
+balance cache for all the processed execution keys.  The cache is
+a map of pubkeys to lamports.
+
+At the start of the block the balance cache is empty.  At the end
+of the block the cache is destroyed.
+
+## Balance Check
+
```
Comment (mvines) [inline book/src/proposals/bankless-leader.md]: What signatures are verified in the Transaction?  Just the execution key's signature or all?

File: book/src/terminology.md [hunk 2]
```diff
@@ -74,6 +74,10 @@ A globally unique identifier that is also a proof that the [entry](terminology.m
 
 The time, i.e. number of [slots](terminology.md#slot), for which a [leader schedule](terminology.md#leader-schedule) is valid.
 
+## execution key
+
+The key in the transaction that pays for the cost of including the transaction in the ledger.  This is the first key in the transaction.
```
Comment (mvines) [inline book/src/terminology.md]: Worth suggesting here that an execution key is highly recommended to be set as Credit-Only?

File: book/src/proposals/bankless-leader.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
+# Bankless Leader
+
+A bankless leader does the minimum amount of work to produce a valid
+block.  The leader is tasked with ingress transactions, sorting and
+filtering valid transactions, arranging them into entries, shredding
+the entries and broadcasting the shreds.
+
+While a validator only needs to reassemble the block and replay
+execution of well formed entries.  For transactions that have
+low execution costs, the leader does 3x more memory operations than
```
[the remark this answers is not in the document]
Comment (aeyakovenko) [reply book/src/proposals/bankless-leader.md]: Both.  Just looking at how transactions flow through the leader, they come unsorted in random packets at the ingress point, then they need to be packed into entries of accepted transactions, then they need to be shredded.  replay stage doesn't need to move any memory besides landing the shreds into the block.

File: book/src/proposals/bankless-leader.md (hunk 1 above)
Comment (aeyakovenko) [reply book/src/proposals/bankless-leader.md]: all are verified

File: book/src/terminology.md (hunk 2 above)
Comment (aeyakovenko) [reply book/src/terminology.md]: @mvines no, its CD to pay for the fee.  it just can't be used in any instruction.

**Comment by mvines** [comment]
Seems like we could back up a sunny day prototype of this and measure the TPS impact pretty easily

File: book/src/terminology.md (hunk 2 above)
Comment (aeyakovenko) [reply book/src/terminology.md]: @mvines, I think it must be CD, otherwise the CO assumptions will fail.

Comment (mvines) [reply book/src/terminology.md]: Yes my bad.

File: book/src/proposals/bankless-leader.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+block.  The leader is tasked with ingress transactions, sorting and
+filtering valid transactions, arranging them into entries, shredding
+the entries and broadcasting the shreds.
+
+While a validator only needs to reassemble the block and replay
+execution of well formed entries.  For transactions that have
+low execution costs, the leader does 3x more memory operations
+before any bank execution than the validator.
+
+## Execution Key
```
Comment (garious) [inline book/src/proposals/bankless-leader.md]: How about we call this a "Fee Account"?

**Comment by garious** [comment]
Seems to me that this would be at most a 2x boost in the leader's Banking Stage, because it allows that stage to load one account where it would otherwise load two. And that assumes that loading accounts is the performance bottleneck where doing simple System transfers (like in bench-tps).  When doing BPF smart contracts (say bench-exchange), I'd expect this would be a much bigger optimization.   If that's the case, that the bench-exchange is significant and bench-tps is insignificant, we should consider relaxing this proposal a bit, and running all System transfers on the leader. By running all System transactions, the user could transfer tokens to other users without it impacting their ability to use that account for subsequent transaction fees.

**Comment by aeyakovenko** [comment]
@garious leader needs to do 2 loads and 2 stores.  With this proposal leader just does 1 load.  so 5x less account work.  The stores might take a bit more time than reads as well.

**Comment by rob-solana** [comment]
the leader isn't exempted from the work of banking in this proposal, the work is merely moved to replay

consecutive leader blocks suffer in this proposal, as ingest has to wait for a bank from replay

**Comment by aeyakovenko** [comment]
@rob-solana Yea, that is true.  But, the current leader can catch up in the replay stage.  

bankless leader + replay stage 

Only does 1 more additional load, which is loading the fee account twice.  But the additional loads should be amortized, since they will remain in memory for the replay stage.

Replay stage is easier to optimize, all the transactions are already in entires.  If its my own leader's block, I can trust that they can all be executed in parallel without the additional check.

**Comment by aeyakovenko** [comment]
> consecutive leader blocks suffer in this proposal, as ingest has to wait for a bank from replay
@rob-solana why is this true?  The next block can be encoded, PoH doesn't depend on bankhash, only the blockhash.

**Comment by rob-solana** [comment]
> @rob-solana Yea, that is true. But, the current leader can catch up in the replay stage.
> 
> bankless leader + replay stage
> 
> Only does 1 more additional load, which is loading the fee account twice. But the additional loads should be amortized, since they will remain in memory for the replay stage.
> 
> Replay stage is easier to optimize, all the transactions are already in entires. If its my own leader's block, I can trust that they can all be executed in parallel without the additional check.

I suppose we can also skip PoH verification, but I still see banking_stage for a new block waiting until all the work of actually banking is finished.  That's ok because the rest of the network is getting their Entries sooner?

**Comment by rob-solana** [comment]
> > consecutive leader blocks suffer in this proposal, as ingest has to wait for a bank from replay
> 
> @rob-solana why is this true? The next block can be encoded, PoH doesn't depend on bankhash, only the blockhash.

the next block depends on _balances_ from the previous block

**Comment by garious** [comment]
Would it make sense to rephrase this proposal as "split BankingStage into FeeVerify+ReplayStage and then put Broadcast before ReplayStage"?  Then the proposal would just be about when and how ReplayStage needs to block FeeVerify for the following slot.

**Comment by aeyakovenko** [comment]
> Would it make sense to rephrase this proposal as "split BankingStage into FeeVerify+ReplayStage and then put Broadcast before ReplayStage"? Then the proposal would just be about when and how ReplayStage needs to block FeeVerify for the following slot.

@garious it doesn't need to block it.  just the cache needs to be consistent with the parent its based on.  as soon as replay is done, the cache can be reset on the next block.

**Comment by aeyakovenko** [comment]
@rob-solana I think I addressed your concerns.  "fee verify" is aware of the base fork its caching, and as soon as replay is done with a new bank it can reset its cache on the next block.  In theory it should lag by at most 2 blocks.

File: book/src/proposals/bankless-leader.md [hunk 5]
```diff
@@ -65,6 +66,19 @@ When the leader finishes the replay stage it can reset the balance
...
 cache which can become active on the next block.
 
+## Reseting the Balance Cache
+
+1. At the start of the block, if the balance cache is uninitialized,
+set the base fork for the balance cache to be the parent of the
+block and create an empty cache.
+
+2. if the cache is initialized, check if block's parents has a new
+finalized bank that is newer than the current base fork for the
```
Comment (rob-solana) [inline book/src/proposals/bankless-leader.md]: frozen bank?

Comment (aeyakovenko) [reply book/src/proposals/bankless-leader.md]: @rob-solana, yea correct.

**Comment by rob-solana** [comment]
yeah, I think my concerns are addressed

**Comment by sakridge** [comment]
> @garious
> 
> > split BankingStage into FeeVerify+ReplayStage
> 
> Replay is asynchronous to FeeVerify, so not sure if that makes sense. But I like the name FeeVerify as the stage. How about “Replace BankingStage with FeeVerify for the leader”

If we are doing FeeVerify *and* replay at the same time most of the time then seems like we are doing even more work than before.

**Comment by garious** [comment]
> If we are doing FeeVerify _and_ replay at the same time most of the time then seems like we are doing even more work than before.

No, that's what Anatoly is saying about "asynchronous".  We just need to watch for places where FeeVerify can't proceed until ReplayStage executes the previous block.

**Comment by aeyakovenko** [comment]
@sakridge two different pipelines which shouldn’t block each other if you have enough cores.

**Comment by sakridge** [comment]
> @sakridge two different pipelines which shouldn’t block each other if you have enough cores.

If we had enough cores, we wouldn't need this design because broadcast wouldn't be impacted. I thought this whole design is to free up cores for broadcast, but now you are saying you need even more cores to run replay and fee verify and broadcast in parallel.

**Comment by aeyakovenko** [comment]
@sakridge that’s not accurate.  If you have 1 pipeline of size N, vs 2  pipelines of size N/2, the latter can achieve better latency with more cores.

**Pull request merged**